### PR TITLE
Create CLI arg for assigning timeout value

### PR DESCRIPTION
**Pull request by chasen-bettinger**
Please fill in the fields below to submit a pull request.  The more information
that is provided, the better.

**Fixes issue #**:
https://github.com/in-toto/in-toto/issues/603

**Description of the changes being introduced by the pull request**:
    - Add `--run-timeout` CLI option to `in-toto-run` that enables user to
    specify a timeout in seconds for subprocess executions.
    - Modify function signature of `execute_link` to include `timeout` and
    function contents which uses that value in order to replace the previous
    behavior of only using the `LINK_CMD_EXEC_TIMEOUT` setting.

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

### CODE REVIEW

File: in_toto/common_args.py [hunk 1]
```diff
@@ -195,6 +195,13 @@ def parse_password_and_prompt_args(args):
     "help": ("generate metadata using dsse (experimental)."),
 }
 
+RUN_TIMEOUT_ARGS = ["--run-timeout"]
+RUN_TIMEOUT_KWARGS = {
+    "dest": "run_timeout",
+    "default": 10,
+    "help": ("float that represents the max timeout for the in-toto-run command"),
```
Comment (adityasaky) [inline in_toto/common_args.py]: This should specify it's in seconds.

File: in_toto/runlib.py [hunk 2]
```diff
@@ -411,6 +410,7 @@ def in_toto_run(
     lstrip_paths=None,
     metadata_directory=None,
     use_dsse=False,
+    timeout=10
```
Comment (adityasaky) [inline in_toto/runlib.py]: Rather than a magic number here, let's retain the default value in settings.py and use it here?

File: in_toto/runlib.py [hunk 3]
```diff
@@ -279,7 +279,7 @@ def _duplicate_streams():
     return proc.poll(), streams["out"], streams["err"]
 
 
-def execute_link(link_cmd_args, record_streams):
+def execute_link(link_cmd_args, record_streams, timeout=10):
```
Comment (adityasaky) [inline in_toto/runlib.py]: We can drop the default value here as the expectation is this is private. AFAICT it's not documented on https://in-toto.readthedocs.io/en/latest/search.html?q=execute_link&check_keywords=yes&area=default

File: in_toto/runlib.py [hunk 4]
```diff
@@ -316,14 +315,14 @@ def execute_link(link_cmd_args, record_streams):
     """
     if record_streams:
         return_code, stdout_str, stderr_str = _subprocess_run_duplicate_streams(
-            link_cmd_args, timeout=float(in_toto.settings.LINK_CMD_EXEC_TIMEOUT)
+            link_cmd_args, timeout=float(timeout)
```
Comment (adityasaky) [inline in_toto/runlib.py]: On a similar note, this function shouldn't be changing the type IMO.

File: in_toto/common_args.py (hunk 1 above)
Comment (adityasaky) [reply in_toto/common_args.py]: And tbh, should we even support sub-second values? I don't know that we have use cases for this value that require such accuracy.

**Review by adityasaky — CHANGES_REQUESTED** [review]
Good start, some initial thoughts.

File: in_toto/common_args.py (hunk 1 above)
Comment (lukpueh) [reply in_toto/common_args.py]: +1 on full seconds

File: tests/rc_test/.in_totorc [hunk 5]
```diff
@@ -1,5 +1,4 @@
 [settings]
 artifact_base_path=r/c/file
 ARTIFACT_EXCLUDE_PATTERNS=r:c:file
-LINK_CMD_EXEC_TIMEOUT=20
```
Comment (lukpueh) [inline tests/rc_test/.in_totorc]: @adityasaky, looks like we forgot to remove this in f88138c90861953c77a1384ea2fcc58126e6fe59.

File: tests/test_runlib.py [hunk 6]
```diff
@@ -469,7 +469,8 @@ def test_record_follow_symlinked_directories(self):
         # Test that all files were recorded including files in linked subdir ...
         self.assertListEqual(
             sorted(list(artifacts_dict.keys())),
-            sorted(self.full_file_path_list + [pair[1] for pair in link_pairs]),
+            sorted(self.full_file_path_list +
+                   [pair[1] for pair in link_pairs]),
```
Comment (lukpueh) [inline tests/test_runlib.py]: This seems unrelated.

File: tests/test_runlib.py [hunk 7]
```diff
@@ -1132,7 +1124,8 @@ def test_create_metadata_with_expected_product(self):
         in_toto_record_start(self.step_name, [], self.key)
         in_toto_record_stop(self.step_name, [self.test_product], self.key)
         link = Metablock.load(self.link_name)
-        self.assertEqual(list(link.signed.products.keys()), [self.test_product])
+        self.assertEqual(list(link.signed.products.keys()),
+                         [self.test_product])
```
Comment (lukpueh) [inline tests/test_runlib.py]: This seems unrelated.

File: doc/source/settings.rst [hunk 8]
```diff
@@ -7,4 +7,4 @@ globals directly in source code or at runtime. This method is discouraged. Inste
 CLI/API arguments should be used.
 
 .. note::
-  The globals ``DEBUG`` and ``LINK_CMD_EXEC_TIMEOUT`` can only be configured directly.
+  The globals ``DEBUG`` can only be configured directly.
```
Comment (lukpueh) [inline doc/source/settings.rst]:
```suggestion
  The global ``DEBUG`` can only be configured directly.
```

File: in_toto/common_args.py [hunk 9]
```diff
@@ -195,6 +195,13 @@ def parse_password_and_prompt_args(args):
     "help": ("generate metadata using dsse (experimental)."),
 }
 
+RUN_TIMEOUT_ARGS = ["--run-timeout"]
+RUN_TIMEOUT_KWARGS = {
+    "dest": "run_timeout",
+    "default": 10,
```
Comment (lukpueh) [inline in_toto/common_args.py]: Let's also use the default from `settings.py` here.

File: tests/rc_test/.in_totorc (hunk 5 above)
Comment (lukpueh) [reply tests/rc_test/.in_totorc]: @chasen-bettinger, you can revert this change and we'll remove the fixture in a separate PR.

**Review by lukpueh — COMMENTED** [review]
Thanks for the contribution, this looks great! Please address @adityasaky's and my comments and we can merge this.

File: tests/test_runlib.py (hunk 6 above)
Comment (lukpueh) [reply tests/test_runlib.py]: ... and breaks the black linter run. Do you maybe have a different max line length setting locally?

File: tests/test_runlib.py (hunk 7 above)
Comment (lukpueh) [reply tests/test_runlib.py]: see https://github.com/in-toto/in-toto/pull/605#discussion_r1211318961

File: tests/test_runlib.py (hunk 6 above)
Comment (chasen-bettinger) [reply tests/test_runlib.py]: Think it was because I was using `autopep8` and not `black`.

**Comment by chasen-bettinger** [comment]
@adityasaky @lukpueh thanks for the review, updated PR accordingly!

File: in_toto/runlib.py [hunk 10]
```diff
@@ -316,14 +315,14 @@ def execute_link(link_cmd_args, record_streams):
...
-            link_cmd_args, timeout=float(in_toto.settings.LINK_CMD_EXEC_TIMEOUT)
+            link_cmd_args, timeout=timeout
         )
 
     else:
         process = subprocess.run(
             link_cmd_args,
             check=False,  # nosec
-            timeout=float(in_toto.settings.LINK_CMD_EXEC_TIMEOUT),
+            timeout=float(timeout),
```
Comment (adityasaky) [inline in_toto/runlib.py]:
```suggestion
            timeout=timeout,
```

File: tests/rc_test/.in_totorc (hunk 5 above)
Comment (adityasaky) [reply tests/rc_test/.in_totorc]: whoops, yeah

**Review by adityasaky — COMMENTED** [review]
This is looking pretty close! Some nits.

**Comment by adityasaky** [comment]
@chasen-bettinger couple of other things:
* there's a mistaken import for the settings module
* missing DCO sign off on the last commit

Also, could you squash / generally clean up your commits? I think we're close to merging this one. :)

**Comment by chasen-bettinger** [comment]
@adityasaky Take a look now.

File: in_toto/common_args.py [hunk 11]
```diff
@@ -195,6 +196,15 @@ def parse_password_and_prompt_args(args):
     "help": ("generate metadata using dsse (experimental)."),
 }
 
+RUN_TIMEOUT_ARGS = ["--run-timeout"]
+RUN_TIMEOUT_KWARGS = {
+    "dest": "run_timeout",
+    "default": LINK_CMD_EXEC_TIMEOUT,
+    "help": (
+        "integer that represents the max timeout in seconds for the in-toto-run command"
```
Comment (adityasaky) [inline in_toto/common_args.py]: Can you include the default value in the help message?

File: in_toto/runlib.py [hunk 12]
```diff
@@ -475,6 +475,9 @@ def in_toto_run(
     use_dsse (optional): A boolean indicating if DSSE should be used to
         generate metadata.
 
+    timeout (optional): An integer indicating the max timeout in seconds
```
Comment (adityasaky) [inline in_toto/runlib.py:478]: Also include the default here.

**Review by adityasaky — APPROVED** [review]
LGTM, final nits from me. Can you amend your commit message? We're exposing the timeout to the user via the CLI and API in addition to the settings file. The intotorc file has been retired from the implementation minus the test artifact so it's not really involved in this PR.

**Comment by adityasaky** [comment]
Also the linter is unhappy.

**Comment by chasen-bettinger** [comment]
@adityasaky Ok, I think I've now made everyone happy.

**Review by lukpueh — APPROVED** [review]
Cheers!

**Comment by adityasaky** [comment]
Thanks for you work on this @chasen-bettinger!
